text/0000-node-api-borrow.md

```diff
@@ -0,0 +1,459 @@
+- Feature Name: Node-API Binary Borrowing
+- Start Date: 2021-03-31
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+
+Neon should provide APIs for _safely_ and _ergonomically_ borrowing binary data. Borrowing should leverage the type system for statically checking safety as much as possible without sacrificing flexibility.
+
+This document proposes an overhaul to the borrow APIs in Neon.
+
+# Motivation
+
+Ahead of the v1.0 release is the best time to make significant changes to the Neon borrowing APIs. Currently, Neon provides these in the `neon::borrow` module. However, with the introduction of [`JsBox`](https://github.com/neon-bindings/rfcs/blob/main/text/0029-jsbox.md), borrowing _only_ applies to buffers. The focused scope allows for more idiomatic wrappers and structuring of the module.
+
+Additionally, there are several safety and correctness related issues with the current `Lock` API that will be addressed. See concerns outlined in the [`JsBox` RFC](https://github.com/neon-bindings/rfcs/blob/main/text/0029-jsbox.md#leverage-existing-borrow-api).
+
+# Reference-level explanation
+
+## Simplified Borrowing
+
+Ideally, Neon would be able to provide statically checked, zero-cost borrowing of JavaScript buffers and views. Fortunately, Neon already provides a type (`Context`) that applies Rust's borrowing rules to the state of the VM.
+
+As long as a reference is held to `Context`, JavaScript cannot execute to invalidate or change a buffer and other native code cannot access the contents.
+
+### Pre-requisites
+
+There are a few methods on the `Context` trait that do not take `&mut self` and can execute JavaScript or access data. These will need a **breaking** change to take an exclusive reference to `self`:
+
+* `Context::lock`
+* `Context::borrow`
+* `Context::borrow_mut`
+* `Context::execute_scoped`
+* `Context::compute_scoped`
+
+### API
+
+The borrowing design consists of a statically checked, ergonomic API and a more complex, but powerful runtime checked API.
+
+#### `trait Borrow`
+
+```rust
+pub trait Borrow: private::Sealed {
```

Comment:
Maybe somewhere this could link to the Rust style guide to explain the "sealed" pattern.